File: color-predictor-service/matrix.py
import random, jsonpickle

import numpy as np


class Matrix:
    def __init__(self, rows, cols):
        self.rows = rows
        self.cols = cols
        self.data = np.zeros((rows, cols))

    def copy(self):
        matrix = Matrix(self.rows, self.cols)
        for i in range(0, self.rows):
            for j in range(0, self.cols):
                matrix.data[i][j] = self.data[i][j]
        return matrix

    @staticmethod
    def fromArray(array):
        return Matrix(len(array), 1).map(lambda *e: array[e[1]])

    @staticmethod
    def subtract(a, b):
        if a.rows != b.rows or a.cols != b.cols:
            return
        return Matrix(a.rows, a.cols).map(lambda _, i, j: a.data[i][j] - b.data[i][j])

    def toArray(self):
        array = []
        for i in range(0, self.rows):
            for j in range(0, self.cols):
                array.append(self.data[i][j])
        return array

    def randomize(self):
        return self.map(lambda *e: random.random() * 2 - 1)

    def add(self, n):
        if isinstance(n, Matrix):
            if self.rows != n.rows or self.cols != n.cols:
                return
            return self.map(lambda e, i, j: e + n.data[i][j])
        return self.map(lambda e: e + n)

    @staticmethod
    def transpose(matrix):
        return Matrix(matrix.cols, matrix.rows).map(lambda _, i, j: matrix.data[j][i])

    @staticmethod
    def multiplyStatic(a, b):
        if a.cols != b.rows:
            return

        matrix = Matrix(a.rows, b.cols)
        for i in range(0, a.rows):
            for j in range(0, b.cols):
                sum = 0
                for idx in range(0, a.cols):
                    sum += a.data[i][idx] * b.data[idx][j]
                matrix.data[i][j] = sum
        return matrix

    def multiply(self, n):
        if isinstance(n, Matrix):
            if self.rows != n.rows or self.cols != n.cols:
                return
            return self.map(lambda e, i, j: e * n.data[i][j])
        return self.map(lambda *e: e[0] * n)

    def map(self, func):
        for i in range(0, self.rows):
            for j in range(0, self.cols):
                val = self.data[i][j]
                self.data[i][j] = func(val, i, j)
        return self

    @staticmethod
    def mapStatic(matrix, func):
        return Matrix(matrix.rows, matrix.cols).map(lambda e, i, j: func(matrix.data[i][j]))

    def print(self):
        print(self.data)

    def serialize(self):
        return jsonpickle.encode(self)

    @staticmethod
    def deserialize(data):
        if isinstance(data, str):
            data = jsonpickle.decode(data)
        matrix = Matrix(data.rows, data.cols)
        matrix.data = data.data
        return matrix

File: websiteWithoutML/js/matrix.js
class Matrix {

    constructor(rows, cols) {
        this.rows = rows
        this.cols = cols
        this.data = Array(this.rows).fill().map(() => Array(this.cols).fill(0))
    }

    copy() {
        let matrix = new Matrix(this.rows, this.cols)
        for (let i = 0; i < this.rows; i++) {
            for (let j = 0; j < this.cols; j++) {
                matrix.data[i][j] = this.data[i][j]
            }
        }
        return matrix
    }

    static fromArray(array) {
        return new Matrix(array.length, 1).map((e, i) => array[i])
    }

    static subtract(a, b) {
        if (a.rows !== b.rows || a.cols !== b.cols) {
            return
        }

        return new Matrix(a.rows, a.cols).map((_, i, j) => a.data[i][j] - b.data[i][j])
    }

    toArray() {
        let array = []
        for (let i = 0; i < this.rows; i++) {
            for (let j = 0; j < this.cols; j++) {
                array.push(this.data[i][j])
            }
        }
        return array
    }

    randomize() {
        return this.map(e => Math.random() * 2 - 1)
    }

    add(n) {
        if (n instanceof Matrix) {
            if (this.rows !== n.rows || this.cols !== n.cols) {
                return
            }
            return this.map((e, i, j) => e + n.data[i][j])
        } else {
            return this.map(e => e + n)
        }
    }

    static transpose(matrix) {
        return new Matrix(matrix.cols, matrix.rows).map((_, i, j) => matrix.data[j][i])
    }

    static multiply(a, b) {
        if (a.cols !== b.rows) {
            return
        }

        return new Matrix(a.rows, b.cols).map((e, i, j) => {
            let sum = 0;
            for (let idx = 0; idx < a.cols; idx++) {
                sum += a.data[i][idx] * b.data[idx][j]
            }
            return sum
        })
    }

    multiply(n) {
        if (n instanceof Matrix) {
            if (this.rows !== n.rows || this.cols !== n.cols)
                return

            return this.map((e, i, j) => e * n.data[i][j])
        } else {
            return this.map((e) => e * n)
        }
    }

    map(func) {
        for (let i = 0; i < this.rows; i++) {
            for (let j = 0; j < this.cols; j++) {
                let val = this.data[i][j]
                this.data[i][j] = func(val, i, j)
            }
        }
        return this
    }

    static map(matrix, func) {
        return new Matrix(matrix.rows, matrix.cols).map((e, i, j) => func(matrix.data[i][j], i, j))
    }

    print() {
        console.table(this.data)
        return this
    }

    serialize() {
        return JSON.stringify(this)
    }

    static deserialize(data) {
        if (typeof data == 'string') {
            data = JSON.parse(data);
        }

        let matrix = new Matrix(data.rows, data.cols)
        matrix.data = data.data
        return matrix
    }

}

if (typeof module !== 'undefinde') {
    module.exports = Matrix
}